mbed-drivers/I2C.h

```diff
@@ -26,11 +26,209 @@
 #include "CThunk.h"
 #include "dma_api.h"
 #include "core-util/FunctionPointer.h"
+#include "core-util/PoolAllocator.h"
 #include "Transaction.h"
 #endif
 
+/**
+ * \file
+ * \brief A generic interface for I2C peripherals
+ *
+ * The I2C class interfaces with an I2C Resource manager in order to initiate Transactions
+ * and receive events. The I2CTransaction class encapsulates all I2C transaction parameters.
+ * The I2CResourceManager class is a generic interface for implementing I2C resource managers.
+ * This will allow for additional classes of I2C device, for example, a bitbanged I2C master.
+ *
+ * I2C Resource managers are instantiated statically and initialized during global init.
+ * There is one Resource Manager per logical port. Logical ports could consist of:
+ *
+ * * Onchip I2C controllers
+ * * I2C Bridges (SPI->I2C bridge, I2C->I2C bridge, etc.)
+ * * Bit banged I2C
+ * * Bit banged I2C over SPI GPIO expander
+ * * More...
+ *
+ * Currently only onchip I2C controllers are supported.
+ *
+ * ## Constructing I2C transactions
+ * To construct an I2C transaction with more than one segment, use the then member,
+ * or the repeated_start member, with a CriticalSectionLock:
+ *
+ * ```C++
+ * void doneCB(bool dir, I2CBuffer buf, int Event) {
+ *     // Do something
+ * }
+ * I2C i2c0(sda, scl);
+ * I2C i2c1(sda, scl);
+ * void app_start (int, char **) {
+ *     uint8_t cmd[2] = {0xaa, 0x55};
+ *     i2c0.transfer_to(addr).tx(cmd,2).then().rx(4).callback(doneCB, I2C_EVENT_ALL);
+ *     // OR
+ *     CriticalSectionLock lock;
+ *     i2c1.transfer_to(addr).tx(cmd,2).repeated_start();
+ *     i2c1.transfer_to(addr).rx(4).callback(doneCB, I2C_EVENT_ALL);
+ * }
+ * ```
+ *
+ * TODO: Use helpers to make it impossible to call rx or tx more than once on the same Transaction
+ */
+
 namespace mbed {
 
+class I2CBuffer {
+public:
+    void set(const Buffer & b) {
+        set(b.buf,b.length);
+    }
+    void set(void * buf, size_t len) {
+        if (len <= sizeof(_packed.data)) {
+            _packed.small = true;
+            _packed.len = len;
+            if (buf) {
+                memcpy(_packed.data, buf, len);
+            }
+        } else {
+            _ref.len = len;
+            _ref.data = buf;
+        }
+    }
+    void * get_buf() {
+        if (_packed.small) {
+            return _packed.data;
+        } else {
+            return _ref.data;
+        }
+    }
+    size_t get_len() const {
+        if (_packed.small) {
+            return _packed.len;
+        } else {
+            return _ref.len;
+        }
+    }
+protected:
+    union {
+        struct {
+            void * data;
+            size_t len;
+        } _ref;
+        struct {
+            uint8_t data[sizeof(_ref)-1];
+            unsigned small:1;
+            unsigned len:7;
+        } _packed;
+    };
```

Comment:
Did you test this on both gcc and armcc? Compilers have been known to do weird stuff with bitfields, especially when packed in an union :)